Sources/UserScript/UserScriptMessaging.swift

```diff
@@ -0,0 +1,263 @@
+            return completion("{}")
+
+        /// Here the client will be expecting a response, so we always to produce one
+        /// We catch errors from handlers so that we can forward the response to clients with the correct context
+        ///
+        /// Most of the logic here is around ensuring we send either `result` or `error` in the response
+        /// Since that's how the Javascript side determines if the request was successful or not.
+        case .respond(let handler, let request):
+            do {
+                try handler(request.params, original) { encodableResponse in
+                    // handle an error
+                    guard let result = encodableResponse else {
+                        let resp = MessageErrorResponse.forRequest(request: request, error: "could not access encodable result").toJSON()
+                        return completion(resp)
+                    }
+
+                    // send the result
+                    if let resp = MessageResponse.toJSON(request: request, result: result) {
+                        return completion(resp)
+                    }
+
+                    let resp = MessageErrorResponse.forRequest(request: request, error: "could not convert result to json").toJSON()
+                    return completion(resp)
+                }
+            } catch let error {
+                let resp = MessageErrorResponse.forRequest(request: request, error: error.localizedDescription).toJSON()
+                return completion(resp)
+            }
+
+        /// We re-throw errors here to let consumers forward them as needed.
+        ///
+        /// Errors here are different to those caught in `.respond` above, because they are not
+        /// always tied to a response. They could be relating the incoming payload being incorrect etc.
+        case .error(let message):
+            let error = NSError(domain: "UserScriptMessaging", code: 0, userInfo: [NSLocalizedDescriptionKey: message])
+            throw error
+        }
+    }
+}
+
+/// An explicit format for specifying how a hostname should be matched
+public enum AllowedOrigin {
+    case etldPlus1(hostname: String)
+    case exact(hostname: String)
+}
+
+/// Force consumers to be explicit about the difference between accepting messages
+/// from 'all domains' vs 'only a subset of domains'
+public enum AllowedOrigins {
+    case all
+    case only(_: [AllowedOrigin])
+
+    public func isAllowed(_ origin: String) -> Bool {
+        switch self {
+        case .all: return true
+        case .only(let allowed):
+            return allowed.contains { allowed in
+                switch allowed {
+                /// exact match
+                case .exact(hostname: let hostname):
+                    return hostname == origin
+                /// etldPlus1, like duckduckgo.com to match dev.duckduckgo.com + duckduckgo.com
+                case .etldPlus1(hostname: let hostname):
+                    return false // todo
```

Comment:
Reminder to take a look at the //todo before merging

Reply:
This one isn't used anywhere yet so I think this can wait (and it's set to false too to be safe)